### ISSUE TITLE
Redirect issue with 3DS enabled

### ISSUE THREAD

**Issue by avaldi** [checkout/checkout-magento2-plugin#78]
Hi guys,

We are facing an issue every time we enable 3DS payments.

Whenever a successful payment is made the following happens:
1. as expected, the user gets redirected to the success page configured in the Checkout Hub (`/success`)
2. right after, the plugin redirects the user again to the 3DS verification page showing an error saying `Merchant.Api.Models.ErrorResponse`
Here you can find the chain of redirects:
![unnamed](https://user-images.githubusercontent.com/1270008/46741814-d5ed1a00-ccb6-11e8-8968-f9b8ab44d07f.png)

I've investigated the issue on my side and it seems like the [PlaceOrderObserver](https://github.com/checkout/checkout-magento2-plugin/blob/bc9344b723b4807342bda499c20040b02da634a4/Observer/PlaceOrderObserver.php#L56) is making the second redirect. Disabling the observer seems to fix the issue, but I'm not sure about possible side-effects.

What's the reason for that observer since anyways the redirect to 3DS happened before?

**Comment by david-fiaty-cko** [comment]
Hi @avaldi,

Here is a demo of the latest version of the module on Magento 2.2.5 similar to your installation:
http://151.80.61.232/m225

It is configured to trigger a 3DS flow with "attempt Non 3D" disabled. If you make a payment test, you will notice that the "Merchant.Api.Models.ErrorResponse" you are having doesn't happen here.

This issue happens when the PHP session is broken or lost. As a matter of fact, if the session doesn't show that the user has already been redirected to the 3DS page, it will happen again. The session doesn't show that the redirection already happened in the cases it's lost or broken.

If it's an issue of broken or lost session then the issue will have to be tackled at a PHP config level.
We have seen this error before with specific server configurations using Redis with cache storage "ON", but it could have another origin.

In this context, disabling the code in the PlaceOrderObserver could have side effects.

Please confirm if you are using a standard configuration for PHP, or if your are using a distributed session handling system with Redis or a similar tool.

If it can help, please send your PHP Info config here or by email.

**Comment by avaldi** [comment]
Hi @david-fiaty-cko 

Thanks for the follow up.
I've been investigating a bit more and it seems like the checkout session is working fine. I checked what's stored in the session *before* the user leaves the webshop (to enter the 3DS page) and *after* he comes back. In both cases the values stored are the same:
<img width="445" alt="screen shot 2018-10-11 at 4 17 33 pm" src="https://user-images.githubusercontent.com/1270008/46803639-9d5e4680-cd71-11e8-9f60-88038919fb4c.png">

Since the `3_d_s_redirect` values is still set in the session, the previously mentioned observer thinks there is a needed redirect and sends the user back to 3DS, which obviously throws an error.

I've checked in the module and it seems like the only place where the `3_d_s_redirect` is unset is in the same [observer](https://github.com/checkout/checkout-magento2-plugin/blob/bc9344b723b4807342bda499c20040b02da634a4/Observer/PlaceOrderObserver.php#L53).

Do you know if there is any other place where the `3_d_s_redirect` should be removed in order not to enter the redirect flow inside the PlaceOrderObserver?

**Comment by david-fiaty-cko** [comment]
@avaldi, if your issue is not at a server level then we should be able to reproduce it on our demo server. We will not be able to change the module code on the master branch as long as we can't reproduce the issue, as many merchants are using it live and don't have the issue reported.

Would you please mind checking the module configuration on our demo server and make sure it's the same as what you have on your server? I will send you admin details by email.

**Comment by david-fiaty-cko** [comment]
@avaldi for information, the 3_d_s_redirect value in session is removed here with an observer:
https://github.com/checkout/checkout-magento2-plugin/blob/master/Observer/PlaceOrderObserver.php

This allows to avoid multiple 3DS redirections and it seems to be working perfectly so far on our demo instance. I could be wrong but I believe you should be looking at server/Redis config issues.

Please have a look here:
https://github.com/magento/magento2/issues/13511

**Comment by david-fiaty-cko** [comment]
@avaldi we are going to run some tests once you have confirmed the config on our demo instance is the same as yours. A phpinfo() for your server config could also be very helpful. Please send it by email if possible.

**Comment by david-fiaty-cko** [comment]
@avaldi there is a new release on the dev branch with support for dynamic 3DS that also disables the redirection in the place order observer. Please let us know if it works better for you.
The new code will be pushed to master early next week as soon as the automated testing process is completed.

**Comment by avaldi** [comment]
Hi @david-fiaty-cko,

I've checked the demo instance you shared with me and it looks like we have pretty much the same configuration. The only main difference is on autocapture time that you have set to `0` while we have it at `0.2`.

The session issue you highlighted (magento/magento2#13511) is not our case as we consistently get redirected, it doesn't happen occasionally. Also, from my test the session is properly maintained, it just contains some wrong values that cause the second 3DS redirect.

I've tested the `dev` branch of this plugin with our configuration and it seems to be working fine for the successful transactions. Let us know as soon as you are able to release it as we will most probably upgrade and start using that version. Meanwhile I'll test the scenario of failed payments (wrong credit card details).
